File: apps/omiro/modules/on-device-ai/src/OnDeviceAIModule.ts
import { requireNativeModule } from 'expo';

// Must stay in sync with the `code` values thrown by OnDeviceAIException in
// OnDeviceAIModule.swift -- that's the only other place these strings are
// defined, so a typo on either side would silently break routing.
export const OnDeviceAIErrorCode = {
  MODEL_UNAVAILABLE: 'MODEL_UNAVAILABLE',
  MISSING_PERMISSION: 'MISSING_PERMISSION',
  GENERATION_FAILED: 'GENERATION_FAILED',
  INVALID_RECURRENCE_RULE: 'INVALID_RECURRENCE_RULE',
  INVALID_DATE_RANGE: 'INVALID_DATE_RANGE',
  CALENDAR_UNAVAILABLE: 'CALENDAR_UNAVAILABLE',
  EVENT_NOT_FOUND: 'EVENT_NOT_FOUND',
  EVENT_READ_ONLY: 'EVENT_READ_ONLY',
  CALENDAR_WRITE_FAILED: 'CALENDAR_WRITE_FAILED',
} as const;

export type CalendarPermissionStatus = 'authorized' | 'denied' | 'notDetermined';

export interface CalendarEventSummary {
  id: string;
  title: string;
  startDate: string;
  endDate: string;
  isAllDay: boolean;
  location: string | null;
  calendarTitle: string | null;
  isEditable: boolean;
}

// Legacy-only view model retained while the unreachable custom event detail
// screen is removed. `listCalendarEventSummaries` never returns these fields.
export interface CalendarEvent extends CalendarEventSummary {
  notes?: string | null;
  participants?: string[];
  recurrenceDescription?: string | null;
}

export interface CalendarDraft {
  title: string;
  startDate: string;
  endDate: string;
  isAllDay: boolean;
  location: string | null;
  notes: string | null;
}

export interface TaskBusyInterval {
  startDate: string;
  endDate: string;
}

export type CalendarEditorResult = 'saved' | 'deleted' | 'cancelled';

export type TimeAssistantResult =
  | { kind: 'answer'; answer: string }
  | {
      kind: 'taskDraft';
      taskTitle: string;
      taskDueAt: string | null;
      taskDurationMinutes: number | null;
      taskScheduledStartAt: string | null;
      taskScheduledEndAt: string | null;
      taskSchedulingWindowStartAt: string | null;
      taskSchedulingWindowEndAt: string | null;
      taskLocation: string | null;
    }
  | { kind: 'availability'; availability: { startDate: string; endDate: string }[] }
  | { kind: 'cancelled' }
  | { kind: 'error'; error: string };

export type TimeProcessingStage = 'understanding' | 'checkingSchedule' | 'preparingSuggestion';

export interface TimeProcessingStageEvent {
  stage: TimeProcessingStage;
  requestToken: string;
}

export type CalendarEventPatch = {
  title?: string;
  startDate?: string;
  endDate?: string;
  location?: string | null;
  notes?: string | null;
};

export type CalendarRecurrenceScope = 'thisEvent' | 'futureEvents';

export interface OnDeviceAIResult {
  text: string;
  isOnDevice: true;
}

// Emitted throughout `askCalendar` -- session start, prompt sent, tool calls
// and results, final response -- so the JS side can render the on-device
// steps as they happen instead of only the final text.
export interface OnDeviceAILogEvent {
  type: string;
  message: string;
  timestamp: number;
  durationMs?: number;
}

export type OnDeviceAIModuleType = {
  getCalendarPermissions(): Promise<CalendarPermissionStatus>;
  requestCalendarPermissions(): Promise<CalendarPermissionStatus>;
  listCalendarEventSummaries(startDate: string, endDate: string): Promise<CalendarEventSummary[]>;
  presentCalendarEvent(id: string): Promise<CalendarEditorResult>;
  presentCalendarDraft(draft: CalendarDraft): Promise<CalendarEditorResult>;
  interpretTimeRequest(
    prompt: string,
    taskBusyIntervals: TaskBusyInterval[],
    requestToken: string,
  ): Promise<TimeAssistantResult>;
  cancelTimeAssistant(requestToken: string): Promise<void>;
  createCalendarEvent(
    title: string,
    startDate: string,
    endDate: string,
    location: string | null,
    recurrenceRule?: string | null,
  ): Promise<CalendarEvent>;
  getCalendarEvent(id: string): Promise<CalendarEvent>;
  updateCalendarEvent(
    id: string,
    patch: CalendarEventPatch,
    recurrenceScope: CalendarRecurrenceScope,
  ): Promise<CalendarEvent>;
  deleteCalendarEvent(id: string, recurrenceScope: CalendarRecurrenceScope): Promise<void>;
  askCalendar(prompt: string): Promise<OnDeviceAIResult>;
  addListener(
    eventName: 'onDeviceAILog' | 'onTimeAssistantStage',
    listener: (event: OnDeviceAILogEvent | TimeProcessingStageEvent) => void,
  ): { remove: () => void };
  // Fires whenever EventKit's store changes, including once a
  // background CalDAV/Exchange sync lands after our first read.
  addListener(eventName: 'onCalendarStoreChanged', listener: () => void): { remove: () => void };
};

export default requireNativeModule<OnDeviceAIModuleType>('OnDeviceAI');

File: apps/omiro/modules/on-device-ai/ios/OnDeviceAIModule.swift
import EventKit
import ExpoModulesCore
import Foundation
import FoundationModels
import os.log

public class OnDeviceAIModule: Module {
  public func definition() -> ModuleDefinition {
    Name("OnDeviceAI")

    Events("onDeviceAILog", "onTimeAssistantStage")

    AsyncFunction("getAvailability") { () async -> String in
      guard #available(iOS 26.0, *) else { return "unsupported" }
      switch SystemLanguageModel.default.availability {
      case .available:
        return "available"
      case .unavailable:
        return "unavailable"
      @unknown default:
        return "unavailable"
      }
    }

    AsyncFunction("getCalendarPermissions") { () async -> String in
      permissionStatusString(EKEventStore.authorizationStatus(for: .event))
    }

    AsyncFunction("requestCalendarPermissions") { () async -> String in
      let status = await requestCalendarAuthorization()
      return permissionStatusString(status)
    }

    AsyncFunction("getCalendarEvents") { (startDate: String, endDate: String) throws -> [[String: Any]] in
      try fetchCalendarEvents(startDate: startDate, endDate: endDate)
    }

    AsyncFunction("listCalendarEventSummaries") { (startDate: String, endDate: String) async throws -> [CalendarEventSummaryRecord] in
      try await MainActor.run {
        try OnDeviceAICalendarCoordinator.shared.summaries(startDate: startDate, endDate: endDate)
      }
    }

    AsyncFunction("presentCalendarEvent") { (id: String) async throws -> String in
      try await OnDeviceAICalendarCoordinator.shared.presentEvent(id: id)
    }

    AsyncFunction("presentCalendarDraft") { (draft: CalendarDraftRecord) async throws -> String in
      try await OnDeviceAICalendarCoordinator.shared.presentDraft(draft)
    }

    AsyncFunction("interpretTimeRequest") { (prompt: String, _ taskBusyIntervals: [TaskBusyIntervalRecord], requestToken: String) async throws -> TimeAssistantResultRecord in
      guard #available(iOS 26.0, *) else {
        var unavailable = TimeAssistantResultRecord()
        unavailable.error = "Natural-language Time requests require Apple Intelligence on this device. You can still browse and edit Calendar manually."
        return unavailable
      }
      let intervals = taskBusyIntervals.compactMap { interval -> TaskBusyInterval? in
        guard let start = iso8601Date(interval.startDate), let end = iso8601Date(interval.endDate), start < end else {
          return nil
        }
        return TaskBusyInterval(startDate: start, endDate: end)
      }
      sendTimeAssistantStage("understanding", requestToken: requestToken)
      let response = try await runTimeAssistant(
        prompt: prompt,
        taskBusyIntervals: intervals,
        requestToken: requestToken,
        onStage: { stage in
          self.sendTimeAssistantStage(stage, requestToken: requestToken)
        }
      )
      return timeAssistantRecord(from: response)
    }

    AsyncFunction("cancelTimeAssistant") { (requestToken: String) in
      Task {
        await cancelTimeAssistant(requestToken: requestToken)
      }
    }

    AsyncFunction("createCalendarEvent") { (
      title: String,
      startDate: String,
      endDate: String,
      location: String?,
      recurrenceRuleValue: String?
    ) throws -> [String: Any] in
      try createCalendarEvent(
        title: title,
        startDate: startDate,
        endDate: endDate,
        location: location,
        recurrenceRuleValue: recurrenceRuleValue
      )
    }

    AsyncFunction("getCalendarEvent") { (id: String) throws -> [String: Any] in
      try fetchCalendarEvent(id: id)
    }

    AsyncFunction("updateCalendarEvent") { (
      id: String,
      patch: [String: Any],
      recurrenceScope: String
    ) throws -> [String: Any] in
      try updateCalendarEvent(id: id, patch: patch, recurrenceScope: recurrenceScope)
    }

    AsyncFunction("deleteCalendarEvent") { (id: String, recurrenceScope: String) throws -> Void in
      try deleteCalendarEvent(id: id, recurrenceScope: recurrenceScope)
    }

    AsyncFunction("askCalendar") { (prompt: String) async throws -> OnDeviceAIResult in
      os_log("askCalendar AsyncFunction invoked", log: onDeviceAILog, type: .info)
      guard #available(iOS 26.0, *) else {
        throw OnDeviceAIException.modelUnavailable
      }

      let startedAt = Date()
      let response = try await runCalendarQuery(
        prompt: prompt,
        onLog: { event, message, durationMs in
          var payload: [String: Any] = [
            "type": event,
            "message": message,
            "timestamp": Date().timeIntervalSince1970 * 1000,
          ]
          if let durationMs {
            payload["durationMs"] = durationMs
          }
          self.sendEvent("onDeviceAILog", payload)
        }
      )
      os_log(
        "askCalendar AsyncFunction resolved after %{public}.0fms",
        log: onDeviceAILog,
        type: .info,
        Date().timeIntervalSince(startedAt) * 1000
      )
      return OnDeviceAIResult(text: response, isOnDevice: true)
    }
  }

  private func sendTimeAssistantStage(_ stage: String, requestToken: String) {
    sendEvent("onTimeAssistantStage", [
      "stage": stage,
      "requestToken": requestToken,
    ])
  }
}

@available(iOS 26.0, *)
private func timeAssistantRecord(from result: TimeAssistantResult) -> TimeAssistantResultRecord {
  let formatter = ISO8601DateFormatter()
  var record = TimeAssistantResultRecord()
  switch result {
  case .answer(let answer):
    record.kind = "answer"
    record.answer = answer
  case .taskDraft(let draft):
    record.kind = "taskDraft"
    record.taskTitle = draft.title
    record.taskDueAt = draft.dueAt.map(formatter.string)
    record.taskDurationMinutes = draft.durationMinutes
    record.taskScheduledStartAt = draft.scheduledStartAt.map(formatter.string)
    record.taskScheduledEndAt = draft.scheduledEndAt.map(formatter.string)
    record.taskSchedulingWindowStartAt = draft.schedulingWindowStartAt.map(formatter.string)
    record.taskSchedulingWindowEndAt = draft.schedulingWindowEndAt.map(formatter.string)
    record.taskLocation = draft.location
  case .availability(let choices):
    record.kind = "availability"
    record.availability = choices.map { choice in
      var value = AvailabilityChoiceRecord()
      value.startDate = formatter.string(from: choice.startDate)
      value.endDate = formatter.string(from: choice.endDate)
      return value
    }
  case .cancelled:
    record.kind = "cancelled"
  case .error(let message):
    record.error = message
  }
  return record
}
